collect_app/src/main/java/org/odk/collect/android/formentry/FormEntryViewModel.java

```diff
@@ -168,6 +180,13 @@ private String getFormIdentifierHash() {
         }
     }
 
+    private void ensureFormController() {
```

Comment:
I'm not sure if I like this...
As I understand it's added to avoid NPE's in cases where we forget to set formController right?
so in collect if we call (in FormEntryActivity):
```
Collect.getInstance().setFormController(formController);
formControllerAvailable(formController);
```
it's not needed. It's needed if we forget the second line?

I'm not sure if it makes a lot of sense. We also might forget to call the first line then: (used in this method)
`Collect.getInstance().getFormController();`
will return null.

so it's still not a perfect way to avoid human mistakes. Additionally it makes the code more coupled.

Reply:
Yeah agreed that it's not full proof and increases coupling. I think really the ViewModels should be able to grab the FormController from some shared component that doesn't just live on `Collect`. I could imagine there being some `FormSessionRepository` or something that lets the ViewModel get the `FormController` lazily (it's loaded if it's not already in memory for instance). I'd be really keen to discuss and try and put the beginnings of something like that (or something better) in #4286.

I do think this is good enough to get out there and fix the crash as the change is small and isolated but yeah, lets not get used to it.

Reply:
As you point out as well it doesn't prevent us making any further mistakes. I think that's a great indicator that we need to find something better.

Reply:
Agreed with all this. @grzesiek2010 I wish we could just set the viewmodels’ formController exactly where we set the App’s but hopefully my explanation makes sense that we can’t without further restructuring because of an extra precondition on formcontrolleravailable. This is definitely meant to be temporary.